### ISSUE TITLE
Naming the scheme, and how the extensibility is going to be used

### ISSUE THREAD

**Issue by kazuho** [kazuho/draft-kazuho-httpbis-priority#86]
We state that the value representation using Structured Headers is "extensible." At the same time, the SETTINGS identifier we define is called "URGENCY." It seems that we think there are two different concepts. I'd like to see if we can agree on what the two concepts are or if we need that separation.

Let me start by asking one question: how do we assume the extensibility to be used in the future?

1) We might assume that we'd only see priority parameter extensions upper-compatible (or coexisting) with the current urgency-based scheme. If that is the case, the SETTINGS identifier should be named EXTENSIBLE or something rather than urgency, as those extensions that might not be urgency-based would also covered by the same SETTINGS identifier.

2) Instead, we might assume that we'd only see priority parameter extensions incompatible with (or exclusive to) the current urgency-based scheme. If that is the case, the SETTINGS identifier should be name URGENCY (as is now), as it is necessary to have the distinction between the SH encoding scheme and the urgency-based prioritization scheme.

My tentative thought is that we are uncertain, though we might be inclined to 1. Then, I think we should go with calling the identifier as EXTENSIBLE. We should also *refrain from* giving a name to the current scheme, as extensions might use some parts of the current mechanism (e.g., the progressive parameter). Having a concept of grouped parameters can become an obstacle. Doing so would give future specifications the most freedom.

I think concrete examples of future extensions (e.g., video streaming use-case) might help us reach conclusion.

**Comment by ianswett** [comment]
I agree we should figure this out.  I was thinking of this more along the lines of the QUIC version and HTTP/3 alpn, where a given version or alpn string indicates a crypto handshake, transport version, and HTTP version at once.  Keeping that parallel, adding a new parameter to the existing SH would be a new enum value.

If the direction really is to add new parameters onto existing ones, then we don't need many SETTINGS values and as you said, it should be called EXTENSIBLE.

There are certainly schemes that don't make much sense with the proposed one(ie: H2).  

Not giving what we have in the draft a name seems awkward to me.

**Comment by LPardue** [comment]
I fully support being able to enhance the behaviour of urgency/progressive with additional parameters. I do wonder exactly what extensibility means though.

The choice of priority scheme is arguably a property of the connection, negotiated via SETTINGS which express a preference ordered list of schemes. Once a scheme is selected it is never changed. How would a browser decide to populate the SETTINGS_PRIORITIES list?

If we continue the line of thinking that a client application directs the local stack prioritization by using a priority header; for any single set of parameters, what priority scheme should the browser negotiate for the client? Is it even too late at that point, a Fetch Request to a first-party resource that uses a set of priority parameters that are incompatible with the selected scheme will result in default prioritization (that's pretty bad right?).  Maybe the browser should open a new connection in such a case? This seems like an area where it is easy to fail to meet developer expectations.

Do we sleep walk into the scenario where the only reasonable thing to do is send all priority parameters for all schemes (aka throw it at the wall and see what sticks method)? Does that work if one priority scheme is active and infers special meaning into the absence of parameters? Some of these problems are related to the discussion on #90, in particular my observation in comment https://github.com/kazuho/draft-kazuho-httpbis-priority/issues/90#issuecomment-555131790. I'm concerned about the ability for actors beyond the client to be able to anaylse and make informed judgments about how to respond given that the behaviour may be determined by a connection-level property that we have no way to articulate.

Consider a stupid scenario where someone defines the SPINAL_TAP scheme, which specifies that the `urgency` level is allowed to go up to 11. A compliant client and server negotiate this scheme just fine and responses are prioritised as expected. Throw in an HTTP/1.1 origin server for collaboration but it understand only URGENCY as defined today. What is it supposed to read into seeing a known parameter with an unexpected value?  What if the alternate scheme just allocated new meaning without renumbering the space? Maybe we should just state "don't ever repurpose another parameter" but there's a fine line with extensibility.

Would it help to make things more explicit and include the prioritization scheme identifier inside the priority header. @kornelski made a case for switching to Structured List on #90 for one form of improvement. After some off-list discussion, I'm inspired to consider an expression change that might address some of the questions on this issue. For example, rather than sending

```
Priority: urgency=6, progressive=?1
```

we send a list that contains the scheme alongside its parameters

```
Priority: urgency;val=6;progressive=?1
```

so the SPINAL_TAP example, which repurposes parameters, becomes unambiguous

```
Priority: spinal_tap;val=6;progressive=?1
```

Such an approach may also allow for versioning within a priority scheme family, without some of the hairyness of ignoring unknown params.

**Comment by ianswett** [comment]
I think the easiest solution to the SPINAL_TAP problem is to avoid redefining a given key/value pair.  So if you want to go to 11, you say:

`Priority: spinal_tap=6;progressive=?1`

**Comment by LPardue** [comment]
I think there are benefits to adding an explicit requirement that the semantic of a parameter MUST not be changed. 

The consequence of that though, based on your candidate solution of adding a new parameter, is that the server receiving this header is getting three parameters - `urgency=<default>` (implicit by ommision), `spinal_tap=6` (explicit), `progressive=?1`. If we simplify the negotiation mechanism (as suggested elsewhere), the client has no way to know which parameter will be used. This also becomes worse when parameters are coupled in some way, the intention of `urgency+progressive` could be very different from `spinal_tap+progressive`. For sanity sake, I would suggest that we introduce guidance that parameters that extend prioritity via substitution MUST NOT infer meaning by combination with other parameters i.e. there is no valid way to express `spinal_tap+progressive`,  instead this should be achieved with `spinal_tap+spinal_tap_progressive`. However, importantly extensions that are additive to a parameter are not subject to these restrictions; for sanity sake we might encourage additive parameters to clearly signal their relationship e.g. an extension for priority ranges should be define a parameter named `urgency-range` rather simply `range`.

**Comment by kazuho** [comment]
While extensibility is a nice-have feature that the CDNs would most likely use, I do not think that it has been the primary goal of the proposal.

IMO, the primary goal has been to provide a scheme that has the following properties:
* a scheme using absolute values (compared to the H2 tree expressing relationships between requests)
* a scheme that can be tuned by the server (it might be debatable if this was the primary goal)
* a scheme that is version-independent

I am happy to see the use of Structured Headers naturally providing extensibility. However, considering the tight schedule, I'd argue that we should *not* emphasize the point that it is extensible, discuss how it can be extended (or expected to be). We can let the extensions find out what to do.

Going back to the topic of the issue, I now think that we should not call the scheme "extensible." Rather, the name should reflect one of the primary goals shown in the above list, or the approach we adopted in order to achieve those goals. For example:
* Absolute Prioritization Scheme for HTTP
* Version-Independent Priorities for HTTP
* Urgency-based Prioritization for HTTP

**Comment by ianswett** [comment]
If we think this scheme is only semi-extensible, does that change our thoughts on how the SETTING should be used?  ie: Might we need something more complex than the original Brad/Lucas priority functionality that was primarily designed to disable HTTP/2 Tree priorities?

**Comment by kazuho** [comment]
> If we think this scheme is only semi-extensible, does that change our thoughts on how the SETTING should be used? ie: Might we need something more complex than the original Brad/Lucas priority functionality that was primarily designed to disable HTTP/2 Tree priorities?

I think it does not change. The fact of *not* clarifying how the new scheme can be extended does not change how it *can* be extended. Because the encoding is the same, and the defined parameters are the same.

**Comment by LPardue** [comment]
After thinking some more, I think one way to look at what we have today is in two parts:

* a new priority signal (header or frame, lets not go down that rabbit hole) that carries some priority information
* urgency-based prioritization that profiles that signal is a specific way

The signal could hold anything an endpoint wanted to send, including some new prioritization scheme. There may be benefit in capturing the scheme in the signal, so that processing it and collaborating on it becomes more stateless. For example, instead of urgency expressed as parameters in a priority dictionary like

```
Priority: u=3, i=?1
```

have a priority list of schemes and their parameters
```
Priority: urgency; u=3; i=?1
```

A client that wants to experiment with a new scheme would send priority under both schemes such as

```
Priority: urgency; u=3; i=?1, spinal_tap; u=11
```

**Comment by kazuho** [comment]
@LPardue While I agree that such a design is possible, I am not sure if we'd prefer doing that.

The semantics of HTTP has extensibility built-in: new header fields. HTTP/2 and HTTP/3 have extensibility built-in: new frames. When a completely different prioritization scheme is designed, they can use those two extension points.

While I like how we use Structured Headers in a way that provides future extensibility, I think that as a secondary goal or something that happened by an accident [1]. And that the scheme we are designing should focus on what it is expected to do.

[1] https://mailarchive.ietf.org/arch/msg/quic/q8DKc7Lj4WuyTlgRj3qXdDMh2dk

**Comment by kornelski** [comment]
In that case, can we use a Structured List where the item is an integer meaning urgency, and takes an optional param `i`.

```
priority: <urgency>
priority: <urgency>; i=<incremental>
```

* For setting the urgency this is the most concise syntax
* urgency + incremental is basically the same
* and it is compatible with the syntax that's needed and most natural for the image acceleration use-case #90, e.g. `priority: 1; i=0; end-at=177, 10; i=1; end-at=5000, 30; i=1`

**Comment by kazuho** [comment]
@kornelski That design has downsides, please refer to the link in my previous comment.

> * and it is compatible with the syntax that's needed and most natural for the image acceleration use-case #90, e.g. `priority: 1; i=0; end-at=177, 10; i=1; end-at=5000, 30; i=1`

It would be compatible *only if* we define the value to be a Structured Headers List. But if we are to define the value as a SH List, then it becomes difficult to define what each list member means.
